### PR TITLE
Fix allocation used storage

### LINKED ISSUE #1050: updateallocation with replace blobber returns `no enough size in allocation`
step to reproduce: 

1. create an allocation with size 2048

2. upload a file of size 1024, then the used size of the allocation an allocation changes to 2048

3. now when you replace a blobber with an other one, I get error: `no enough size in allocation`. 

when you replace the blobber, the used_size of the allocation changes to `1536`. So there is 512 bytes of storage available. Even though there is 512 bytes of data available in the allocation, we still get the above error. this is because we check against the actual file size 

meaning, issue only happens when we try to replace a blobber in an allocation when remaining size is 0

https://0chain.slack.com/archives/C02AV6MKT36/p1687115793572579?thread_ts=1686935002.932209&cid=C02AV6MKT36

### PR DESCRIPTION
**Pull request by Jayashsatolia403**
## Tests
Tasks to complete before merging PR:
- [ ]  Ensure system tests are passing. If not [Run them manually](https://github.com/0chain/0chain/actions/workflows/system_tests.yml) to check for any regressions :clipboard:
- [ ]  Do any new system tests need added to test this change? do any existing system tests need updated? If so create a PR at [0chain/system_test](https://github.com/0chain/system_test)
- [ ]  Merge your system tests PR to master AFTER merging this PR

## Associated PRs (Link as appropriate):
- blobber:
- gosdk:
- system_test:
- zboxcli:
- zwalletcli:
- Other: ...

### CODE REVIEW

File: code/go/0chain.net/smartcontract/dbs/event/allocation.go [hunk 1]
```diff
@@ -286,11 +286,18 @@ func (edb *EventDb) updateAllocationsStats(allocs []Allocation) error {
...
+		AddUpdate("moved_to_challenge", movedToChallengeList, "allocations.moved_to_challenge").
+		AddUpdate("moved_back", movedBackList, "allocations.moved_back").
+		AddUpdate("write_pool", writePoolList, "allocations.write_pool").Exec(edb).Error
+
+	//return CreateBuilder("allocations", "allocation_id", allocationIdList).
+	//	AddUpdate("used_size", usedSizeList, "allocations.used_size + t.used_size").
+	//	AddUpdate("num_writes", numWritesList, "allocations.num_writes + t.num_writes").
+	//	AddUpdate("moved_to_challenge", movedToChallengeList, "allocations.moved_to_challenge + t.moved_to_challenge").
+	//	AddUpdate("moved_back", movedBackList, "allocations.moved_back + t.moved_back").
+	//	AddUpdate("write_pool", writePoolList, "allocations.write_pool - t.moved_to_challenge + t.moved_back").Exec(edb).Error
```
Comment (Kishan-Dhakan) [inline code/go/0chain.net/smartcontract/dbs/event/allocation.go]: remove commented code

**Review by Kishan-Dhakan — CHANGES_REQUESTED** [review]

File: code/go/0chain.net/smartcontract/storagesc/allocation.go [hunk 2]
```diff
@@ -1765,17 +1765,17 @@ func (sc *StorageSmartContract) finishAllocation(
...
+		MovedBack:        allocation.MovedBack,
+		WritePool:        allocation.WritePool,
 	}
 
-	if w.Size > 0 {
-		alloc.MovedToChallenge = movedTokens
-	} else if w.Size < 0 {
-		alloc.MovedBack = movedTokens
-	}
+	logging.Logger.Info("emitUpdateAllocationStatEvent", zap.Any("alloc", alloc))
```
Comment (dabasov) [inline code/go/0chain.net/smartcontract/storagesc/allocation.go]: do not use any please

File: code/go/0chain.net/smartcontract/storagesc/allocation.go [hunk 3]
```diff
@@ -1765,17 +1765,17 @@ func (sc *StorageSmartContract) finishAllocation(
...
-		UsedSize:     w.Size,
+		AllocationID:     allocation.ID,
+		UsedSize:         allocation.Stats.UsedSize,
+		NumWrites:        allocation.Stats.NumWrites,
+		MovedToChallenge: allocation.MovedToChallenge,
+		MovedBack:        allocation.MovedBack,
+		WritePool:        allocation.WritePool,
 	}
 
-	if w.Size > 0 {
```
Comment (dabasov) [inline code/go/0chain.net/smartcontract/storagesc/allocation.go:1774]: what was it for?

File: code/go/0chain.net/smartcontract/dbs/event/allocation.go [hunk 4]
```diff
@@ -236,12 +236,12 @@ func mergeUpdateAllocStatsEvents() *eventsMergerImpl[Allocation] {
 }
 
 func mergeAllocationStatsEvents() *eventsMergerImpl[Allocation] {
-	return newEventsMerger[Allocation](TagUpdateAllocationStat, withAllocStatsMerged())
+	return newEventsMerger[Allocation](TagUpdateAllocationStat, withUniqueEventOverwrite())
 }
 
 func withAllocStatsMerged() eventMergeMiddleware {
```
Comment (dabasov) [inline code/go/0chain.net/smartcontract/dbs/event/allocation.go]: do we need this middleware?

**Comment by service-0chain** [comment]
Manual system tests [cancelled] with the following config
| config  | value |
| ------------- | ------------- |
| tests skipped  | false  |
| system-tests  | sprint-june-5  |
| gosdk  | fix/alloc-size  |
| zboxcli  | sprint-june-5  |
| zwalletcli  | sprint-june-5  |
| 0chain  | fix/size  |
| blobber  | staging  |
| authorizer  | staging  |
| 0box  | staging  |
| 0dns  | staging  |

**Comment by service-0chain** [comment]
Manual system tests [failure] with the following config
| config  | value |
| ------------- | ------------- |
| tests skipped  | false  |
| system-tests  | sprint-june-5  |
| gosdk  | sprint-june-5  |
| zboxcli  | sprint-june-5  |
| zwalletcli  | sprint-june-5  |
| 0chain  | fix/size  |
| blobber  | staging  |
| authorizer  | staging  |
| 0box  | staging  |
| 0dns  | staging  |

**Comment by service-0chain** [comment]
Manual system tests [success] with the following config
| config  | value |
| ------------- | ------------- |
| tests skipped  | false  |
| system-tests  | sprint-june-5  |
| gosdk  | sprint-june-5  |
| zboxcli  | sprint-june-5  |
| zwalletcli  | sprint-june-5  |
| 0chain  | fix/size  |
| blobber  | staging  |
| authorizer  | staging  |
| 0box  | staging  |
| 0dns  | staging  |

**Comment by service-0chain** [comment]
Manual system tests [cancelled] with the following config
| config  | value |
| ------------- | ------------- |
| tests skipped  | false  |
| system-tests  | sprint-june-5  |
| gosdk  | sprint-june-5  |
| zboxcli  | fix/alloc-size  |
| zwalletcli  | sprint-june-5  |
| 0chain  | fix/size  |
| blobber  | staging  |
| authorizer  | staging  |
| 0box  | staging  |
| 0dns  | staging  |

**Comment by service-0chain** [comment]
Manual system tests [failure] with the following config
| config  | value |
| ------------- | ------------- |
| tests skipped  | false  |
| system-tests  | sprint-june-5  |
| gosdk  | fix/alloc-size  |
| zboxcli  | fix/alloc-size  |
| zwalletcli  | sprint-june-5  |
| 0chain  | fix/size  |
| blobber  | staging  |
| authorizer  | staging  |
| 0box  | staging  |
| 0dns  | staging  |

**Comment by service-0chain** [comment]
Manual system tests [success] with the following config
| config  | value |
| ------------- | ------------- |
| tests skipped  | false  |
| system-tests  | sprint-june-5  |
| gosdk  | fix/alloc-size  |
| zboxcli  | fix/alloc-size  |
| zwalletcli  | sprint-june-5  |
| 0chain  | fix/size  |
| blobber  | staging  |
| authorizer  | staging  |
| 0box  | staging  |
| 0dns  | staging  |

**Comment by service-0chain** [comment]
Manual system tests [cancelled] with the following config
| config  | value |
| ------------- | ------------- |
| tests skipped  | false  |
| system-tests  | sprint-june-5  |
| gosdk  | sprint-june-5  |
| zboxcli  | sprint-june-5  |
| zwalletcli  | sprint-june-5  |
| 0chain  | fix/size  |
| blobber  | staging  |
| authorizer  | staging  |
| 0box  | staging  |
| 0dns  | staging  |

**Review by dabasov — APPROVED** [review]
lgtm

**Review by Kishan-Dhakan — APPROVED** [review]

**Pull request merged**